compiler/rustc_ast/src/token.rs

```diff
@@ -42,11 +42,87 @@ pub enum BinOpToken {
     Shr,
 }
 
+// This type must not implement `Hash` due to the unusual `PartialEq` impl below.
+#[derive(Copy, Clone, Debug, Encodable, Decodable, HashStable_Generic)]
+pub enum InvisibleOrigin {
+    // From the expansion of a metavariable in a declarative macro.
+    MetaVar(MetaVarKind),
+
+    // Converted from `proc_macro::Delimiter` in
+    // `proc_macro::Delimiter::to_internal`, i.e. returned by a proc macro.
+    ProcMacro,
+
+    // Converted from `TokenKind::Interpolated` in
+    // `TokenStream::flatten_token`. Treated similarly to `ProcMacro`.
+    FlattenToken,
```

Comment:
Is there any observable difference between `ProcMacro` and `FlattenToken`?
(It's possible that I already asked this on the older PRs.)

Reply:
There is no difference. The final commit in #124141 removes `FlattenToken` and flattening in general.